ml_metadata/tools/mlmd_bench/thread_runner.cc

```diff
@@ -0,0 +1,114 @@
+// failed.
+// The thread runner will first loops over all the executable workloads in
+// benchmark and executes them one by one. Each workload will have a
+// `thread_stats_list` to record the stats of each thread when executing the
+// current workload.
+// During the execution, each operation will has a `op_stats` to record current
+// operation statistic. Each `op_stats` will be used to update the
+// `thread_stats`.
+// After the each thread has finished the execution, the workload stats will be
+// generated by merging all the thread stats inside the `thread_stats_list`. The
+// performance of the workload will be reported according to the workload stats.
+tensorflow::Status ThreadRunner::Run(Benchmark& benchmark) {
+  for (auto& workload : benchmark.workloads()) {
+    Stats thread_stats_list[num_threads_];
+    // `workload.second` is the number of operations for current executable 
+    // workload.
+    int64 op_per_thread = workload.second / num_threads_;
+    std::unique_ptr<MetadataStore> set_up_store;
+    TF_RETURN_IF_ERROR(CreateMetadataStore(mlmd_config_, &set_up_store));
+    // `workload.first` is the executable workload.
+    TF_RETURN_IF_ERROR(workload.first->SetUp(set_up_store.get()));
+    {
+      // Create a thread pool for multi-thread execution.
+      tensorflow::thread::ThreadPool pool(tensorflow::Env::Default(),
+                                          "mlmd_bench", num_threads_);
+      // `total_done` is used for reporting progress along the way.
+      int64 total_done = 0;
+      int64 thread_index = 0;
+      for (int64 i = 0; i < num_threads_; ++i) {
+        pool.Schedule([&]() {
+          // Uses a lock to ensure the consistence of concurrent reading and
+          // writing of `thread`.
+          lock_.Lock();
+          int64 curr_thread_index = thread_index;
+          thread_index++;
+          lock_.Unlock();
+          // Each thread uses a different MLMD client instance to talk to the
+          // same back-end.
+          std::unique_ptr<MetadataStore> store;
+          CreateMetadataStore(mlmd_config_, &store);
+          int64 start_index = op_per_thread * curr_thread_index;
+          thread_stats_list[curr_thread_index].Start();
+          // Executes the current workload by the specified `work_items_index`.
+          int64 work_items_index = start_index;
+          while (work_items_index < start_index + op_per_thread) {
+            // Each operation has a op_stats.
+            OpStats op_stats;
+            tensorflow::Status status =
+                workload.first->RunOp(work_items_index, store.get(), op_stats);
+            // Handles abort issues for concurrent writing to the db.
+            if (!status.ok()) {
+              // Retry the current operation.
+              continue;
+            }
+            work_items_index++;
+            lock_.Lock();
+            total_done++;
+            lock_.Unlock();
```

Comment:
qq, this looks a barrier will affect all threads and serialize all functions, what're other alternatives to do the cmd reporting?

Reply:
Yes, we can discuss about it.